producer/data/ref_counted.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package data
+
+import (
+	"sync"
+
+	"github.com/m3db/m3msg/producer"
+
+	"go.uber.org/atomic"
+)
+
+// OnFinalizeFn will be called when the data is being finalized.
+type OnFinalizeFn func(d producer.RefCountedData)
+
+type refCountedData struct {
+	sync.RWMutex
+	producer.Data
+
+	onFinalizeFn OnFinalizeFn
+
+	refCount            *atomic.Int32
+	isDroppedOrConsumed bool
+}
+
+// NewRefCountedData creates RefCountedData.
+func NewRefCountedData(data producer.Data, fn OnFinalizeFn) producer.RefCountedData {
+	return &refCountedData{
+		Data:                data,
+		refCount:            atomic.NewInt32(0),
+		onFinalizeFn:        fn,
+		isDroppedOrConsumed: false,
+	}
+}
+
+func (d *refCountedData) Filter(fn producer.FilterFunc) bool {
+	return fn(d.Data)
+}
+
+func (d *refCountedData) IncRef() {
+	d.refCount.Inc()
+}
+
+func (d *refCountedData) DecRef() {
+	rc := d.refCount.Dec()
+	if rc == 0 {
+		d.finalize(producer.Consumed)
+	}
+	if rc < 0 {
+		panic("invalid ref count")
+	}
+}
+
+func (d *refCountedData) IncReads() {
+	d.RLock()
+}
+
+func (d *refCountedData) DecReads() {
+	d.RUnlock()
+}
+
+func (d *refCountedData) Bytes() []byte {
+	return d.Data.Bytes()
+}
+
+func (d *refCountedData) Size() uint64 {
+	return uint64(d.Data.Size())
+}
+
+func (d *refCountedData) Drop() {
+	d.finalize(producer.Dropped)
+}
+
+func (d *refCountedData) IsDroppedOrConsumed() bool {
+	d.RLock()
+	r := d.isDroppedOrConsumed
+	d.RUnlock()
+	return r
+}
+
+func (d *refCountedData) finalize(r producer.DataFinalizeReason) {
+	d.Lock()
+	if d.isDroppedOrConsumed {
+		d.Unlock()
+		return
+	}
+	d.isDroppedOrConsumed = true
+	if d.onFinalizeFn != nil {
+		d.onFinalizeFn(d)
```

Comment:
I'm still a bit confused about the difference between the data finalizer and the refcounted data finalizer. Can you give an example how they might be used?

Reply:
The only use case right now for data finalizer is to decrease the size of the buffer, so that whenever a message is fully consumed the buffer size will be updated. This way the size of the buffer is kept up to date and avoids the cases where `Buffer` call thinks the buffer is full and iterate through the list to reduce size for already consumed data.

Reply:
Right makes sense, what's an example of refcounted data finalizer?

Reply:
You mean the `Data.Finalize()` I assume? That is up to the user, most likely returning the underlying bytes to the pool and get notification about the `FinalizeReason` (Dropped or Consumed) and handle accordingly